File: consys-core/src/main/scala/de/tuda/stg/consys/core/store/Ref.scala
package de.tuda.stg.consys.core.store

import scala.language.higherKinds

/**
 * Created on 10.12.19.
 *
 * @author Mirko Köhler
 */
trait Ref[StoreType <: Store, T <: StoreType#ObjType] {
	def resolve(tx : StoreType#TxContext) : StoreType#HandlerType[T]
}

File: consys-japi/src/main/java/de/tuda/stg/consys/japi/Ref.java
package de.tuda.stg.consys.japi;

/**
 * Created on 27.01.20.
 *
 * @author Mirko Köhler
 */
public interface Ref<T> {

	default T ref() {
		throw new UnsupportedOperationException("Use the consys-compiler plugin to resolve calls to ref().");
	}

	<R> R getField(String fieldName);

	<R> void setField(String fieldName, R value);

	<R> R invoke(String methodName, Object... args);
}
